core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java

```diff
@@ -0,0 +1,231 @@
+                case IS_DISTINCT_FROM -> or(
+                        new IsNullPredicate(argument),
+                        new NotExpression(between(argument, argumentType, calculateRangeStartInclusive(right, argumentType), calculateRangeEndInclusive(right, argumentType))));
+                case LESS_THAN -> new ComparisonExpression(LESS_THAN, argument, toExpression(calculateRangeStartInclusive(right, argumentType), argumentType));
+                case LESS_THAN_OR_EQUAL -> new ComparisonExpression(LESS_THAN_OR_EQUAL, argument, toExpression(calculateRangeEndInclusive(right, argumentType), argumentType));
+                case GREATER_THAN -> new ComparisonExpression(GREATER_THAN, argument, toExpression(calculateRangeEndInclusive(right, argumentType), argumentType));
+                case GREATER_THAN_OR_EQUAL -> new ComparisonExpression(GREATER_THAN_OR_EQUAL, argument, toExpression(calculateRangeStartInclusive(right, argumentType), argumentType));
+            };
+        }
+
+        private Object calculateRangeStartInclusive(Object yearVariable, Type type)
+        {
+            int year = toIntExact((Long) yearVariable);
+            if (type == DATE) {
+                LocalDate firstDay = LocalDate.ofYearDay(year, 1).with(firstDayOfYear());
+                return firstDay.toEpochDay();
+            }
+            if (type instanceof TimestampType timestampType) {
+                if (timestampType.isShort()) {
+                    LocalDateTime dateTime = LocalDateTime.MIN.withYear(year).with(firstDayOfYear());
+                    return dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND + LongMath.divide(dateTime.getNano(), NANOSECONDS_PER_MICROSECOND, UNNECESSARY);
+                }
+                LocalDateTime dateTime = LocalDateTime.MIN.withYear(year).with(firstDayOfYear());
+                long endInclusiveMicros = dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND
+                        + LongMath.divide(dateTime.getNano(), NANOSECONDS_PER_MICROSECOND, UNNECESSARY);
+                return new LongTimestamp(endInclusiveMicros, 0);
+            }
+            throw new UnsupportedOperationException("Unsupported type: " + type);
+        }
+
+        private Object calculateRangeEndInclusive(Object yearVariable, Type type)
+        {
+            int year = toIntExact((Long) yearVariable);
+            if (type == DATE) {
+                LocalDate lastDay = LocalDate.ofYearDay(year, 1).with(lastDayOfYear());
+                return lastDay.toEpochDay();
+            }
+            if (type instanceof TimestampType timestampType) {
+                if (timestampType.isShort()) {
+                    LocalDateTime dateTime = LocalDateTime.of(year, 12, 31, 23, 59, 59).with(lastDayOfYear());
+                    return dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND + 999999;
```

Comment:
`999999` is a magic number - let's rather define a constant

Reply:
Changed the logic to avoid such constant.